### PR TITLE
feat(rest): allow assigning express settings

### LINKED ISSUE #2420: How to configure qs
I understand that LB is using `qs` under the hood to parse query string. Is there a way, through binding or other mean, to configure qs as it only parses object for a depth of 5. The rest is kept stingified (see https://github.com/ljharb/qs#parsing-objects) ?

### PR DESCRIPTION
**Pull request by ludohenin**
Add a new rest configuration for `queryParser`

Fixes #2420 

## Checklist

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

### CODE REVIEW

**Comment by ludohenin** [comment]
@raymondfeng not sure what to test here. I've made this PR to start the discussion.

File: packages/rest/src/rest.server.ts [hunk 1]
```diff
@@ -937,6 +946,7 @@ export interface RestServerOptions {
   openApiSpec?: OpenApiSpecOptions;
   apiExplorer?: ApiExplorerOptions;
   sequence?: Constructor<SequenceHandler>;
+  queryParser?: qs.IParseOptions;
```
Comment (raymondfeng) [inline packages/rest/src/rest.server.ts]: I suggest that we extend it for all express related settings - see http://expressjs.com/en/4x/api.html#app.set.

```ts
expressSettings?: {[name: string]: any};
```

Comment (ludohenin) [reply packages/rest/src/rest.server.ts]: Sound good. question though:

1. do you want to set the query parser custom function from the config?
1. do we really need support for `jsonp`, `views`, `view engine` and `view cache`?

Comment (raymondfeng) [reply packages/rest/src/rest.server.ts]: > 1. do you want to set the query parser custom function from the config?

IMO, it should be possible.

> 2. do we really need support for `jsonp`, `views`, `view engine` and `view cache`?

We can skip such options if they don't make sense to LB4. The other option is to take them as-is if they don't do any harm.

Comment (ludohenin) [reply packages/rest/src/rest.server.ts]: Second try ... I think it's actually fine to leave it up to the app developper to pass whatever he/she wants as settings. Is it more inline with what you had in mind ?

*note: will update the doc once you're ok with the implementation*

File: packages/rest/src/rest.server.ts [hunk 2]
```diff
@@ -937,6 +951,7 @@ export interface RestServerOptions {
   openApiSpec?: OpenApiSpecOptions;
   apiExplorer?: ApiExplorerOptions;
   sequence?: Constructor<SequenceHandler>;
+  expressSettings?: {[name: string]: any};
```
Comment (raymondfeng) [inline packages/rest/src/rest.server.ts]: Please add `// tslint:disable-next-line:no-any` above `expressSettings`.

File: packages/rest/src/rest.server.ts [hunk 3]
```diff
@@ -234,6 +234,20 @@ export class RestServer extends Context implements Server, HttpServerLike {
...
   }
 
+  /**
+   * Apply express settings. Bind functions to the server context (not express).
+   */
+  protected _applyExpressSettings() {
+    Object.entries(this.config.expressSettings || {}).forEach(
+      ([key, value]) => {
+        if (typeof value === 'function') {
+          value = value.bind(this);
```
Comment (raymondfeng) [inline packages/rest/src/rest.server.ts]: Why `bind` here? I think it's the responsibility of the code that provides the configuration to supply the correct function.

File: docs/site/Server.md [hunk 4]
```diff
@@ -211,6 +211,38 @@ export async function main() {
...
+### Configure query parser
+
+`qs` is used to parse query string with the default package settings.
+
+You can override the defaults by setting your own configuration:
+
+```ts
+// Rest server config
+rest: {
+  queryParser: {...},
```
Comment (raymondfeng) [inline docs/site/Server.md]: Please fix the option here to be `expressSettings`.

**Comment by raymondfeng** [comment]
To unit test the `expressSettings`, you can create a subclass of `RestServer`, which can then access `_expressApp` and use `_expressApp.get(...)` to verify.

File: docs/site/Server.md (hunk 4 above)
Comment (ludohenin) [reply docs/site/Server.md]: done

File: packages/rest/src/rest.server.ts (hunk 3 above)
Comment (ludohenin) [reply packages/rest/src/rest.server.ts]: not sure, I was thinking it's better than having express bound to this. Removed it.

File: packages/rest/src/rest.server.ts (hunk 2 above)
Comment (ludohenin) [reply packages/rest/src/rest.server.ts]: done

**Comment by ludohenin** [comment]
@raymondfeng should be closer now

**Comment by raymondfeng** [comment]
@ludohenin Please run `npm run lint:fix` to fix style/formatting [issues reported by CI](https://travis-ci.org/strongloop/loopback-next/builds/495203052?utm_source=github_status&utm_medium=notification).

**Comment by ludohenin** [comment]
@raymondfeng done but there's still an error with node 10

**Comment by raymondfeng** [comment]
@ludohenin Thanks. The CI failure is not related to your change.

**Comment by raymondfeng** [comment]
@ludohenin I rebased your branch against latest master. Can you grant me the permission per https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/ so that I can push such changes to the PR?

File: packages/rest/src/rest.server.ts (hunk 3 above)
Comment (raymondfeng) [reply packages/rest/src/rest.server.ts]: BTW, we need to keep the default `'query parser': 'extended'` if not present. I fixed it in my clone (to be pushed back to the PR)

**Comment by ludohenin** [comment]
@raymondfeng should be good now

File: packages/rest/src/rest.server.ts [hunk 5]
```diff
@@ -214,7 +214,7 @@ export class RestServer extends Context implements Server, HttpServerLike {
   protected _setupRequestHandlerIfNeeded() {
     if (this._expressApp) return;
     this._expressApp = express();
-    this._expressApp.set('query parser', 'extended');
+    this._applyExpressSettings();
```
Comment (bajtos) [inline packages/rest/src/rest.server.ts]: If I am reading these lines correctly, then:

- before the change, we were explicitly configuring the Express instance to use `extended` query parser
- now we use the default `query parser` configuration as provided by Express

I see that Express uses `extended` query parser by default, so I guess this is ok.

https://expressjs.com/en/4x/api.html#app.settings.table

**Pull request merged**